ACF Block Support

rsm0128 opened the pull request:
For ACF blocks, `acf` property is added.

- Added ACF Gutenberg Block support.
- Added video support for Gallery ACF type.
- Fixed bug for the blocks with empty DOM

rsm0128 commented:
@dChiamp This PR adds ACF block support. For sliders block, we can create an ACF field group with a gallery field and assign it to custom ACF block. Please review and let me know if you have any questions.

- ACF field group screenshot
![image](https://github.com/user-attachments/assets/336f98df-d80d-47b6-9083-f1b7a0d0fa5a)

- Post editor:
![image](https://github.com/user-attachments/assets/83969bb3-f4e5-4566-8d8b-c6a05682d469)

- API response:
![image](https://github.com/user-attachments/assets/8bfe253e-170c-4b06-9c16-2fcfb469ac6b)

drewbaker commented:
@rsm0128 How do you define a block is available? Meaning, what makes it show up in the locations drop down? 

From memory, you need to define the blocks in PHP, is that still the case?
https://www.advancedcustomfields.com/resources/create-your-first-acf-block/#create-functionsphp

Would be so good if there was a way to define the blocks in WordPress via GUI?

We will never want our custom blocks to have custom UI in Gutenberg. We are OK with the custom blocks just using the ACF input UIs. I think that is the `preview` parameter in the PHP function, but I may be wrong.

rsm0128 commented:
@drewbaker Yes, of course, we can create a panel to create ACF blocks dynamically if we don't need custom UI for the blocks. Shall we add the function to the fuxt-backend?

drewbaker commented:
Where do you think would be best to put that? As an extension to ACF settings somewhere perhaps? Or as a settings page to fuxt-api?

@dChiamp got any thoughts?

dChiamp commented:
What do you guys think is the easiest / fastest? @rsm0128 how did you set that up in the example? is there a way we can temporarily add that to gnet so we can build the new slideshow vue component in the meantime?

rsm0128 commented:
Here is the code to create an empty ACF block. The block.php and block.css files can be empty files

```
function fuxt_custom_blocks() {

    // Abort early if custom blocks not supported (not ACF Pro).
    if( !function_exists('acf_register_block_type') ) {
		return;
    }

    // Slideshow block.
    acf_register_block_type(array(
        'name'              => 'slideshow',
        'title'             => 'Slideshow',
        'description'       => 'A custom slideshow block.',
        'category'          => 'media',
        'mode'              => 'edit',
        'keywords' 			=> array('image', 'slideshow'),
        'icon'				=> 'slides',
        'render_template'   => get_template_directory() . '/blocks/slideshow/block.php',
        'enqueue_style' 	=> get_template_directory_uri() . '/blocks/slideshow/block.css',
    ));
}
add_action('acf/init', 'fuxt_custom_blocks');

```

rsm0128 commented:
@dChiamp FYI, you can find an example in `/org-funkhaus/bvla-api/wp-content/themes/bvla-backend/functions/acf-custom-blocks.php`

dChiamp commented:
@rsm0128 added this plugin version from the PR to gnet, and the `/acf-custom-blocks.php` file, and empty files in `/blocks` dir, and loaded in functions file in the theme. then added the metafield and assigned it to slideshow block like in the screenshot above, but this is what it's returning:

![image](https://github.com/user-attachments/assets/d23fb2d6-3b4e-42d0-a45b-fa7118bd6f62)

Any idea what we are missing?

rsm0128 commented:
@dChiamp I investigated gnet and identified the causes of the problem:
- There were two `fuxt-api` plugin installations (v0.1.2 and v0.1.3) on the site, with v0.1.2 activated.
- The [slideshow ACF field](https://gnet.flywheelsites.com/wp-admin/post.php?post=574&action=edit)'s returnFormat was set to `Image ID` when the block was added to the [Fallout 76 + Fallout 4](https://gnet.flywheelsites.com/wp-admin/post.php?post=571&action=edit) page. However, it was set to `Image Array` during my check, causing a null value return.

![image](https://github.com/user-attachments/assets/f18da54d-dcd7-435a-932d-3f66a8064f9d)

I have addressed these issues, and now the ACF block data returns as expected.
![image](https://github.com/user-attachments/assets/3ccf48a2-0c23-4194-bd3f-38ce7315c43e)

FYI, the ACF data is added as a new property `acf` while leaving `attrs->data` unchanged. Do you think it would be better to update the `attrs->data` field instead of adding new `acf` property?
Please let me know your preference.
cc @drewbaker

lux-v commented:
@rsm0128 @dChiamp @drewbaker  I created a new custom gutenberg block [statistics](https://gnet.flywheelsites.com/wp-admin/post.php?post=838&action=edit) which uses `Repeater` as a field type. 

I realized it's returning `null` values for fields that are nested inside repeater(image below)
![image](https://github.com/user-attachments/assets/7b7d2789-4228-4f8d-aac7-7e0035b66841)


I found a bug and pushed a [fix](https://github.com/funkhaus/fuxt-api/pull/74/commits/4f4999ce15bfd7a4b118a9db2856a49347132ea5) to this PR.
![image](https://github.com/user-attachments/assets/67d84652-9aa5-4a67-a2bd-d067399fa384)

dChiamp commented:
Hey @rsm0128 we ran into another unexpected result with acf and custom blocks. In this example you can see the value of acf always returns the first instance of the block:

<img width="456" alt="acf-custom-gutenberg" src="https://github.com/user-attachments/assets/fd8cefa5-d10a-406a-a4f2-d68e0b5947ed" />

that should return `{icon: FourBars, heading: out of home}`

rsm0128 commented:
Hi @dChiamp OK, Let me check and fix the problem

rsm0128 commented:
@dChiamp, @lux-v I fixed the issue in the latest commits. Please check and let me know if you are still facing the issue.
